docs/proposals/20230123-azure-service-operator.md

```diff
@@ -0,0 +1,235 @@
+  verbs:
+  - get
+  - list
+  - watch
+  - create
+  - update
+  - delete
+```
+The above permissions may overlap with ASO resources already installed by a user not intended to be managed by CAPZ.
+
+### Risks and Mitigations
+
+The most immediate risks with adopting ASO are regressions of various forms. Firstly, Azure API calls are core to CAPZ's functionality, so even small behavioral differences may cause issues preventing Azure resources from being reconciled as expected. Next, the modified reconciliation flow may cause operations against workload cluster to take longer to complete, in particular if it takes more reconciliations to perform a task. Finally, integration with ASO may introduce new or worsen existing bottlenecks limiting how many CAPZ resources can be managed. Scaling bottlenecks might be Azure API rate limits or management cluster Kubernetes API server load because of the increased number of Kubernetes resources. These risks will be monitored closely as described in the [Test Plan](#test-plan). Any significant regressions found can be rolled back backwards-compatibly by reverting the relevant changes.
+
+Another shorter-term risk is that although ASO v2 is [officially supported by Microsoft](https://azure.github.io/azure-service-operator/introduction/frequently-asked-questions/#what-is-the-support-model), it has not yet published a GA release. Until then, ASO may make breaking changes between pre-releases. The remaining work to be completed before ASO v2's first GA release is being tracked here: https://github.com/Azure/azure-service-operator/milestone/20.
+
+Further down the road, there is an apparent increased risk that CAPZ's lifespan may extend beyond that of its Azure interface when ASO is used instead of the API or SDK directly, as ASO has not yet been proven as a mission-critical interface to Azure to the same extent as Azure's REST API and Go SDK.
+
+Since ASO can be freely installed by users, there is also a risk that a CAPZ-installed ASO may conflict with a user-installed instance. Before CAPZ installs ASO, it should verify that ASO is already either not installed or installed and configured as expected by CAPZ.
+
+The difference between CAPZ's and ASO's metrics stories may also be considered a risk. CAPZ's metrics currently include traces that measure how long each step in a reconciliation takes, including all Azure API calls. Once those equivalent API calls are being done from ASO instead of CAPZ, CAPZ's metrics will lose that data. However, that lost data may no longer be meaningful to CAPZ since Azure API calls will be happening out-of-band with CAPZ reconciliations. ASO does measure how many Azure API calls are made and how long each takes and exposes those in the Prometheus format, along with other metrics as documented at https://azure.github.io/azure-service-operator/introduction/metrics/. However, users relying on the current structure of metrics will likely experience breaking changes.
+
```

Comment:
Is there a potential risk, albeit probably minor, to release/feature implementation velocity if something needs to be implemented in ASO first, wait for its release cycle, and then implemented in CAPZ? I'm not sure what the planned release cycle will be once it goes GA for implementing the new features available from the SDK.

Reply:
Definitely, I'll add that here. I agree that it's probably minor, considering I don't think CAPZ is often blocked on adding new features because no SDK version exists yet with that feature. I'm also hopeful that CAPZ can have a decent amount of influence on when and what ASO releases.

Reply:
@matthchr Does ASO rely on the SDK to have an API version published before it can be generated for ASO to enable new features? Or does that depend only on the new API spec? 

I'm thinking if there's still only one "hop" (i.e. cycle of asking for some API version to be generated) from "API version available in Azure" to "ready to use by CAPZ's dependency" then this may not actually be all that different from how CAPZ currently relies on the SDK. Still definitely worth noting here, but I'm wondering if I should phrase the updated dependency chain from CAPZ -> SDK as

CAPZ -> ASO

or

CAPZ -> ASO -> SDK

Reply:
No, it doesn't. Not sure how much you know about Azure SDKs but basically, they are code-generated from a public OpenAPI/Swagger specification with a custom code-generator called [AutoRest](https://github.com/Azure/autorest). ASO also is a code-generator that also uses the OpenAPI/Swagger specification, so there is no usage of the SDKs in ASO* and we don't need to wait for the SDKs to exist to release a resource in ASO. 

With that said, there are practical concerns that at least today make ASO somewhat slower than the SDKs to release. They are:
1. ASO just hasn't hit the economies of scale that the SDKs have yet. We're getting there as we add more new RPs/resources but I would say in general right now we're slower on average.
2. ASO is like the CLI rather than like SDKs in that it ships all together (CRDs + controller) rather than individual packages like the SDKs. This means even if we're done with resources A, B, C, we may be waiting for D, E, F before next release. This isn't the case with SDKs they ship individually.
3. Some resources require some manual massaging to make great experiences with in k8s. Examples include: being able to export certain fields (think connection details) to a configMap for import by a pod. SDks don't have to worry about this, ASO we spend a bit of time making sure we get those things right so it's more involved than just "generate the CRDs and we're done"
4. Some Azure resources are very hard to model well in k8s (this is pretty rare). The best example is resources that are infinitely recursive. No problem for the SDKs, but you can't model an infinitely recursive resource in a k8s CRD at all.

*That's actually not quite true because there's some handcrafted code in a few places that uses the SDKs, and we use the Track2 internal pipeline implementation to send/receive requests to Azure. 95% we don't use the SDKs though and most resources do not use the SDKs except for the shared pipeline

Reply:
This is excellent, thanks! I'll include all these points in the doc.